Parallelize Tests in CI (+ Refactor and Improvements)

aminya opened the pull request:
### Description of the Change
This:
- On Windows, runs core renderer tests using two Azure agents. This in addition to double the overall test speed, will allow rerunning the failed ones in case some fail due to timeouts.
- allows requesting any tests on any operating system
- parallelizes core renderer tests by using separate child processes for each file. 
- allows rerunning only the failed tests rather than restarting the whole CI.
- reduces Windows timeouts by running the tests on windows-2019.


### Verification
The CI passes

### Release Notes
Same as the description

### Closes
This supersedes these old pull requests: 
#18270

aminya committed:
parallelize core tests

aminya committed:
allow requesting parallel tests for all OS

aminya committed:
run tests using async.parallel

aminya committed:
don't use temp.track()

cleanup process fails when the file is locked (in parallel tests)

aminya committed:
warn before error reporting

Helps visually

aminya committed:
test: use Azure format for printing

aminya committed:
run package tests in parallel for windows tests

aminya committed:
Run windows renderer tests in parallel

aminya committed:
Run windows core main tests in the build step

aminya committed:
always upload atom windows.zip for x64

aminya committed:
bootstrap in case cache misses

aminya committed:
macos: run core main tests in the build phase

aminya committed:
print the used testCommand for failed tests

aminya committed:
update message about finding a single application to run the tests

aminya committed:
fix: rename windows job back to Windows

aminya committed:
collect console log in renderer tests

Print the output finally to ease diagnosing

aminya committed:
Use async.series

Revert b61475ffe420672b3360b50aba7b0f347ea319d4

DeeDeeG commented:
We could re-write this PR to not include the heavy use of Azure DevOps templates. Which would make for smaller and more targeted changes, and be easier to review.

Let us know; We'd be very glad to see this improvement land at this repo, so CI can run faster here.

aminya commented:
> We could re-write this PR to not include the heavy use of Azure DevOps templates.

I don't think rewriting this is a good idea. They should merge #21081 before this. Not using templates just makes a soup of code that it is hard to maintain.

aminya committed:
function to create packageTestSuites only when needed

aminya committed:
function to create coreRenderProcessTestSuites

only when needed

aminya committed:
refactor spawnTest

aminya committed:
collect output only if stdout exits

Duv4lC4ruso commented on script/vsts/platforms/linux.yml:
```diff
@@ -30,12 +30,26 @@ jobs:
         inputs:
-          keyfile: 'package.json, script/vsts/platforms/linux.yml, **/package-lock.json, !**/node_modules/**/package-lock.json, !**/.*/**/package-lock.json'
-          targetfolder: '**/node_modules, !**/node_modules/**/node_modules'
-          vstsFeed: 'bae1bc26-220d-43c7-a955-4de039370de2'
```
#19441 
#10326 
#20511

aminya replied:
Please remove the spam comment which references incorrect issues. If I report your account that is much worse @Duv4lC4ruso.

DeeDeeG replied:
I've already reported it as spam. I don't see any particularly legitimate activity on the account. Sorry if it's a real person, but the account activity looks like a bot.

aminya committed:
rename windows jobs 

to be consistent with macOS

sadick254 commented:
@aminya I think this needs rebasing.

aminya commented:
@sadick254 This is ready to go! 😉

sadick254 reviewed (commented):
@aminya Thank you for your contribution. Looking at this PR 2 questions come to mind.
- Does this affect the overall speed of running `script/test`
- What is the impact of spawning a `child_process` for every spec file?

Another thing i have noted is, this PR changes the behaviour of running `script/test`. This PR makes it mandatory to specify the suite one would like to run. Initially `script/test` would run all the core tests without extra `args`.

aminya commented:
> @aminya Thank you for your contribution. Looking at this PR 2 questions come to mind.
> 
> * Does this affect the overall speed of running `script/test`
> * What is the impact of spawning a `child_process` for every spec file?
> 

Thank you for asking. I haven't noticed any negative impact on performance. The speed bottleneck of running the tests is not in spawning the child_process (which might take less than 1 second), but in running the tests themselves. 

Using separate child_process allows us to selectively choose what we want to run, and rerun them if they fail (#21295). Using this I have reduced the overall performance and test pass rate.

With that being said there is still room for improvement. If you have noticed we use `async.series` which is not ideal. I have been mostly concerned with increasing the overall test experience. To maximize the speed of running these, we should apply some other changes and change this to `async.parallel`. I can do that after we merged #21295.

> Another thing i have noted is, this PR changes the behaviour of running `script/test`. This PR makes it mandatory to specify the suite one would like to run. Initially `script/test` would run all the core tests without extra `args`.

Yes, and I think this is a better approach. However, if you don't like this, it has an easy fix. Here instead of throwing an error, we can run all the tests:
https://github.com/atom/atom/pull/21109/files#diff-0e2bc0f03fe54e5e0fff129205e6ee92R294

aminya committed:
refactor the static parts out of the test function which runs by async

aminya committed:
test on windows-2019

This increases the test-pass rate and reduced the timeouts

https://github.com/atom-ide-community/atom/pull/109

sadick254 commented:
@aminya Thanks for the clarification. 
> Yes, and I think this is a better approach. However, if you don't like this, it has an easy fix. Here instead of throwing an error, we can run all the tests

No need. I think this is ok. 👍
